ksqldb-rest-app/src/test/java/io/confluent/ksql/api/integration/ScalablePushBandwidthThrottleIntegrationTest.java

```diff
@@ -0,0 +1,213 @@
+
+    RestIntegrationTestUtil.createStream(REST_APP, TEST_DATA_PROVIDER);
+
+    makeKsqlRequest("CREATE TABLE " + AGG_TABLE + " AS "
+        + "SELECT PAGEID, LATEST_BY_OFFSET(USERID) AS USERID FROM " + TEST_STREAM + " GROUP BY PAGEID;"
+    );
+  }
+
+  @AfterClass
+  public static void classTearDown() {
+    REST_APP.getPersistentQueries().forEach(str -> makeKsqlRequest("TERMINATE " + str + ";"));
+  }
+
+  @Before
+  public void setUp() {
+    vertx = Vertx.vertx();
+    restClient = REST_APP.buildKsqlClient();
+  }
+
+  @After
+  public void tearDown() {
+    if (vertx != null) {
+      vertx.close();
+    }
+    REST_APP.getServiceContext().close();
+  }
+
+  @SuppressFBWarnings({"DLS_DEAD_LOCAL_STORE"})
+  @Test
+  public void scalablePushBandwidthThrottleTest() {
+    assertAllPersistentQueriesRunning();
+    String veryLong = createDataSize(100000);
+    final CompletableFuture<StreamedRow> header = new CompletableFuture<>();
+    final CompletableFuture<List<StreamedRow>> complete = new CompletableFuture<>();
+    String sql = "SELECT CONCAT(\'"+ veryLong + "\') as placeholder from " + AGG_TABLE + " EMIT CHANGES LIMIT 1;";
+
+    // scalable push query should succeed 10 times
+    for (int i = 0; i < 11; i += 1) {
+      makeRequestAndSetupSubscriber(sql,
+          ImmutableMap.of("auto.offset.reset", "latest"),
+          header, complete );
+      TEST_HARNESS.produceRows(TEST_TOPIC, TEST_DATA_PROVIDER, FormatFactory.JSON, FormatFactory.JSON);
+      System.out.println(i);
+    }
+
+    // scalable push query should fail on 11th try since it exceeds 1MB bandwidth limit
+      try {
+        makeQueryRequest(sql,
+            ImmutableMap.of("auto.offset.reset", "latest"));
+        TEST_HARNESS.produceRows(TEST_TOPIC, TEST_DATA_PROVIDER, FormatFactory.JSON, FormatFactory.JSON);
+
+        throw new AssertionError("New scalable push query should have exceeded bandwidth limit ");
+      } catch (KsqlException e) {
+        assertThat(e.getMessage(), is(RATE_LIMIT_MESSAGE));
+      }
+  }
+
+  private static String createDataSize(int msgSize) {
+    StringBuilder sb = new StringBuilder(msgSize);
+    for (int i=0; i<msgSize; i++) {
+      sb.append('a');
+    }
+    return sb.toString();
+  }
+
+  private static void makeKsqlRequest(final String sql) {
+    RestIntegrationTestUtil.makeKsqlRequest(REST_APP, sql);
+  }
+
+  private void makeRequestAndSetupSubscriber(
+      final String sql,
+      final Map<String, ?> properties,
+      final CompletableFuture<StreamedRow> header,
+      final CompletableFuture<List<StreamedRow>> future
+  ) {
+    publisher = makeQueryRequest(sql, properties);
+    subscriber = new QueryStreamSubscriber(publisher.getContext(), future, header);
+    publisher.subscribe(subscriber);
+  }
+
+  StreamPublisher<StreamedRow> makeQueryRequest(
+      final String sql,
+      final Map<String, ?> properties
+  ) {
+    final RestResponse<StreamPublisher<StreamedRow>> res =
```

Comment:
It would be good to at least also test the `/query-stream` endpoint is exactly the same way since that's a slightly different code path and is used by the java client.  You may need to expose the async version of the method in the KsqlRestClient to do that.  `KsqlTarget.postQueryRequestStreamedAsync` uses that endpoint and can be exposed in `KsqlRestClient`

Reply:
Discussed offline and agreed to add this http2 test in a follow up in order to make release cut deadline